In this project, we will be using a dataset from Kaggle (scraped by the user "orgesleka"), which contains data from 50,000 auto sale transactions during a particular time in Germany conducted on eBay. 

The data has been "dirtied" so that it will be useful as a cleaning and analysis exercise. We will be cleaning this data by removing outlier entries for a variety of reasons, and looking at the results to see what we can learn about the German used car market.

Below, we will import our needed libraries (NumPy and pandas) as well as import the dataset .csv file into our environment to begin our work.

In [126]:
#Importing libraries

import numpy as np
import pandas as pd

#Reading the csv file and assigning to a variable. UTF-8 was giving errors, so Latin-1 was used instead.

autos = pd.read_csv("autos.csv",encoding="Latin-1")

In [127]:
#Using Jupyter Notebooks built-in function of rendering head and tail values of the variableby entering it to preview
#our results.

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [128]:
#Continuing our first look at our data set.

autos.info()
print("")
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

What we can see from our initial look will be some main tasks in cleaning this dataset:

- The columns "odometer" and "price" could be relabled and cast as numerical values rather than strings.

- The columns "dateCrawled", "lastSeen", "dateCreated" could be cast into datetime objects rather than strings.

- We have a significant amount of NaN values in multiple columns. We will be leaving these in for now, as these columns don't seem relevant to the analysis we are conducting. 

- The columns are using "camelcase", we will want to change this to "snakecase".

Some lesser items, which may or may not need to be done:

- We will likely want to translate our data from German to English to aid our interpretation.

- There are car model names that will need to be changed to their actual name. ("7er" is not a model of BMW, likely meaning some other 7 series.)

In [129]:
#Changing the column names from camelcase to snakecase, and then comparing the results.

#Importing a fresh copy of the dataset so that this block can be run independently in Jupyter Notebook

before_columns = pd.read_csv("autos.csv",encoding="Latin-1").columns
print("BEFORE:")
print(before_columns)

#The renaming of the columns

autos.rename(columns={'yearOfRegistration':"registration_year","monthOfRegistration":"registration_month",
              "notRepairedDamage":"unrepaired_damage","dateCreated":"ad_created","offerType":"offer_type",
             "vehicleType":"vehicle_type","powerPS":"power_ps","fuelType":"fuel_type","nrOfPictures":"nr_of_pictures",
              "postalCode":"postal_code","lastSeen":"last_seen","dateCrawled":"date_crawled"},inplace=True)


#Printing the resulting column names as well as the head of the updated dataframe for comparison

print("")
print("AFTER:")
print(autos.columns)
print("")
print(autos.head())

BEFORE:
Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

AFTER:
Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58

The reason for changing to snakecase is that snakecase is the expected casing structure in python. We may encounter issues if the environment is expecting a snakecase style input, when the input is actually camelcase.

Now that column names have been cleaned, we can look at the data further to see other things we would like to clean, such as turning numeric data into a numeric data type instead of a string, and dropping redundant columns where all the values are the same.

In [130]:
#Gives us statistics about our data, to see what we want to work on

autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Going further in our data cleaning, we can see some other tasks to complete:

The "seller" and "offer_type" columns have only two different unique values, and one of those values occurs 49999 times, and the other occurs once. These two columns can be removed, as they don't provide any insight into our data.

For the same reason as above, the "nr_of_pictures" column is all zeros, and can be removed as well.

Columns like "odometer" and "price" can be turned from string to numeric, and then have their column names adjusted to reflect the units in which the column values are being measured in.

Columns like "abtest", "gearbox", and "unrepaired_damage" may look redundant since they only have 2 unique values, but the frequency number suggests enough of a distribution between the two values that it is worth keeping in our dataset.

The columns with date and time information (date_crawled, ad_created) could be cast as datetime objects to get data such as how long these posts took to be viewed after they were created, to see if certain types of cars are "hotter" than others based on how quickly users viewed listings of certain types of cars.

First, we will address the "odometer" and "price" columns as mentioned above.

In [131]:
#Casting all of the values as ints once the characters
#that prevent this are removed ("km", "$" and "",""). Then, renaming each column to include the unit value.

autos["odometer"] = autos["odometer"].str.replace("km","",regex=True)
autos["odometer"] = autos["odometer"].str.replace(",","",regex=True)
autos["odometer"] = autos["odometer"].astype(int)
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

autos["price"] = autos["price"].str.replace("$","",regex=True)
autos["price"] = autos["price"].str.replace(",","",regex=True)
autos["price"] = autos["price"].astype(int)
autos.rename({"price":"price_usd"},axis=1,inplace=True)

#Checking to see our results for the column name, dytpe, and values in the column.

print(autos.dtypes)
autos

date_crawled          object
name                  object
seller                object
offer_type            object
price_usd              int32
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int32
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
nr_of_pictures         int64
postal_code            int64
last_seen             object
dtype: object


date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price_usd   abtest vehicle_type  registration_year  \
0      privat    Angebot       5000  control          bus               2004   
1      privat    Angebot       8500  control    limousine               1997   
2      privat    Angebot       8990     test    limousine               2009   
3      privat    Angebot       4350  control   kleinwagen               2007   
4      privat    Angebot       1350     test        kombi               2003   
...       ...        ...        ...      ...          ...                ...   
49995  privat    Angebot      24900  control    limousine               2011   
49996  privat    Angebot       1980  control       cabrio               1996   
49997  privat    Angebot      13200     test       cabrio               2014   
49998  privat    Angebot      22900  control        kombi               2013   
49999  privat    Angebot       1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere       150000                   3       lpg   
1      automatik       286     7er       150000                   6    benzin   
2        manuell       102    golf        70000                   7    benzin   
3      automatik        71  fortwo        70000                   6    benzin   
4        manuell         0   focus       150000                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5       100000                   1    diesel   
49996    manuell        75   astra       150000                   5    benzin   
49997  automatik        69     500         5000                  11    benzin   
49998    manuell       150      a3        40000                  11    diesel   
49999    manuell       101  vectra       150000                   1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-0

After completing and checking that cleaning task, we will now move on to see if there are any outliers in our data that we may want to remove, and if so, remove them. We will start with the "odometer_km" and "price_usd" columns we just worked with.

In [132]:
#Number of unique values

print("odometer_km number of unique values: " + str(autos["odometer_km"].unique().shape[0]))
print("price_usd number of unique values: " + str(autos["price_usd"].unique().shape[0]))

#Printing the statistical distribution of their values

print("")
print("odometer_km statistics:")
print(autos["odometer_km"].describe())
print("")
print("price_usd statistics:")
print("")
print(autos["price_usd"].describe().apply(lambda x: format(x, 'f'))) # .apply() function to suppress scientific notation

odometer_km number of unique values: 13
price_usd number of unique values: 2357

odometer_km statistics:
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

price_usd statistics:

count       50000.000000
mean         9840.043760
std        481104.380500
min             0.000000
25%          1100.000000
50%          2950.000000
75%          7200.000000
max      99999999.000000
Name: price_usd, dtype: object


In [133]:
#Counts of each value for odometer_km

print("")
print("odometer_km value counts:")
print(autos["odometer_km"].value_counts())

#Largest value counts for price_usd, as well as largest price amounts
#We are curating these results since there are so many unique values

print("")
print("price_usd value counts ascending:")
print(autos["price_usd"].value_counts().head(30).sort_index(ascending=True))
print("")
print("price_usd highest prices:")
print(autos["price_usd"].nlargest(30))


odometer_km value counts:
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

price_usd value counts ascending:
0       1421
300      384
350      335
500      781
550      356
600      531
650      419
700      395
750      433
800      498
850      410
900      420
950      379
999      434
1000     639
1100     376
1200     639
1250     335
1300     371
1500     734
1600     327
1800     355
1999     322
2000     460
2200     382
2500     643
3000     365
3500     498
4500     394
5500     340
Name: price_usd, dtype: int64

price_usd highest prices:
39705    99999999
42221    27322222
27371    12345678
39377    12345678
47598    12345678
2897     11111111
24384    11111111
11137    10000000
47634     3890000
7814      1300000
22947     1234566
514        999999
43049      999999
37585      999990

There is only 13 different values in the "odometer_km" column, and the distribution seems normal, the only problem being all values 150,000 and greater are entered in the dataset as 150,000. Since this is nearly 2/3 of the dataset, it would harm our conclusions to remove these, so we will leave them in.

In the "price_usd" column we can see an abnormal amount of 0 values, and knowing people don't sell cars for free, we will eliminate the rows containing 0 values for price. Also, when looking at the highest values, there are 14 values of 999,990 or greater, which we will be removing, as 999,990 is nearly 3x the amount of the next highest value of 350,000.

In [134]:
#Removing values of 0 and greater than 999990
autos = autos.loc[(autos["price_usd"]>0) & (autos["price_usd"]<999990),:]

#Checking our work to see that those values have been removed.
print("")
print("price_usd value counts ascending:")
print(autos["price_usd"].value_counts().head(30).sort_index(ascending=True))
print("")
print("price_usd highest prices:")
print(autos["price_usd"].nlargest(30))
print("")
print(autos.describe())


price_usd value counts ascending:
300     384
350     335
400     321
500     781
550     356
600     531
650     419
700     395
750     433
800     498
850     410
900     420
950     379
999     434
1000    639
1100    376
1200    639
1250    335
1300    371
1500    734
1600    327
1800    355
1999    322
2000    460
2200    382
2500    643
3000    365
3500    498
4500    394
5500    340
Name: price_usd, dtype: int64

price_usd highest prices:
36818    350000
14715    345000
34723    299000
35923    295000
12682    265000
47337    259000
38299    250000
37840    220000
40918    198000
43668    197000
28090    194000
20351    190000
17140    180000
11433    175000
32840    169999
18509    169000
22673    163991
45387    163500
10500    155000
33638    151990
49668    145000
32185    139997
2454     137999
14268    135000
49815    130000
1878     129000
8232     128000
2751     120000
44406    120000
43282    119900
Name: price_usd, dtype: int32

           price_usd  registration_ye

We can conclude the data removed were outliers by seeing the standard deviation of the "price_usd" column reduce from 481104.3805 to 9059.854754. Since standard deviation is an indication of variance in the data set, we can see by a rough estimate, the variance of the data was reduced to about 1/53 of what it was, by only removing about 3% of the rows in the dataset. 

Seeing less than 3% of the dataset cause such a huge impact on the standard deviation, as well as having practical cause for eliminating that data suggests this is the correct move.

Next, we will be taking a closer look at the columns containing date data to understand the date range the data covers, starting with the "date_crawled", "ad_created", and "last_seen" columns.

In [135]:
#Seeing what the values look like
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We can see the values are in the YYYY-MM-DD HH:MM:SS format. Since we are just concerned with dates, we can select the first 10 characters of each string, which will contain the date. We will do this accross each of the three columns and analyze them one at a time.

In [136]:
#Capturing the date range, and what percentage the occurence of each date is of the total range of dates.
date_crawled_dates = autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

#Formatting a cleaned up view of the date range and percentages, converting the decimal value to actual percentage.
print("date_crawled_dates, as percentage of total:")
print(str(date_crawled_dates * 100)+"%")

date_crawled_dates, as percentage of total:
2016-03-05    2.532688
2016-03-06    1.404304
2016-03-07    3.601359
2016-03-08    3.329558
2016-03-09    3.308967
2016-03-10    3.218367
2016-03-11    3.257490
2016-03-12    3.691959
2016-03-13    1.566972
2016-03-14    3.654896
2016-03-15    3.428395
2016-03-16    2.960980
2016-03-17    3.162772
2016-03-18    1.291053
2016-03-19    3.477813
2016-03-20    3.788737
2016-03-21    3.737259
2016-03-22    3.298672
2016-03-23    3.222485
2016-03-24    2.934212
2016-03-25    3.160712
2016-03-26    3.220426
2016-03-27    3.109235
2016-03-28    3.486050
2016-03-29    3.409863
2016-03-30    3.368681
2016-03-31    3.183363
2016-04-01    3.368681
2016-04-02    3.547823
2016-04-03    3.860805
2016-04-04    3.648718
2016-04-05    1.309585
2016-04-06    0.317101
2016-04-07    0.140019
Name: date_crawled, dtype: float64%


We can see from the above, that the user who captured this data set had scraped the data over a 33 day period from March 5th to April 7th of 2016. The distribution of posts crawled by date is fairly uniform, with each date accounting for between .14% and 3.86% of the total.

In [137]:
#Capturing the date range, and what percentage the occurence of each date is of the total range of dates.
ad_created_dates = autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

#Formatting a cleaned up view of the date range and percentages, converting the decimal value to actual percentage.
#Since this range is wider, we will also take a closer look at the head/tail due to the truncated printing of the 76 values.
print("ad_created_dates, as percentage of total:")
print(str(ad_created_dates * 100)+"%")
print("")
print("first 15 dates:")
print(str(ad_created_dates.head(35) * 100)+"%")
print("")
print("latest 15 dates:")
print(str(ad_created_dates.tail(50) * 100)+"%")

ad_created_dates, as percentage of total:
2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
                ...   
2016-04-03    3.885514
2016-04-04    3.685782
2016-04-05    1.181921
2016-04-06    0.325337
2016-04-07    0.125605
Name: ad_created, Length: 76, dtype: float64%

first 15 dates:
2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
2015-12-30    0.002059
2016-01-03    0.002059
2016-01-07    0.002059
2016-01-10    0.004118
2016-01-13    0.002059
2016-01-14    0.002059
2016-01-16    0.002059
2016-01-22    0.002059
2016-01-27    0.006177
2016-01-29    0.002059
2016-02-01    0.002059
2016-02-02    0.004118
2016-02-05    0.004118
2016-02-07    0.002059
2016-02-08    0.002059
2016-02-09    0.002059
2016-02-11    0.002059
2016-02-12    0.004118
2016-02-14    0.004118
2016-02-16    0.002059
2016-02-17    0.002059
2016-02-18    0.004118
2016-02-19    0.00617

As far as when the ads were created, the period is much larger. The range of dates from when the first post in the dataset was created to the latest post is 301 days. (From June 11th, 2015 to April 7th, 2016.)

Another thing we can see is that although the earliest date is June 11th, 2015, the date crawled range of March 5th 2016 to April 7th 2016 contains nearly all the occurences, with about 95% of the total data entries falling between that range.

In [138]:
#Capturing the date range, and what percentage the occurence of each date is of the total range of dates.
last_seen_dates = autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

#Formatting a cleaned up view of the date range and percentages, converting the decimal value to actual percentage.
print("last_seen_dates, as percentage of total:")
print(str(last_seen_dates * 100)+"%")

last_seen_dates, as percentage of total:
2016-03-05     0.107073
2016-03-06     0.432410
2016-03-07     0.539483
2016-03-08     0.741275
2016-03-09     0.959539
2016-03-10     1.066612
2016-03-11     1.237517
2016-03-12     2.378256
2016-03-13     0.889529
2016-03-14     1.260167
2016-03-15     1.587563
2016-03-16     1.645218
2016-03-17     2.808607
2016-03-18     0.735097
2016-03-19     1.583445
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-22     2.137342
2016-03-23     1.853186
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-26     1.680222
2016-03-27     1.564913
2016-03-28     2.085864
2016-03-29     2.234119
2016-03-30     2.477093
2016-03-31     2.378256
2016-04-01     2.279419
2016-04-02     2.491506
2016-04-03     2.520334
2016-04-04     2.448265
2016-04-05    12.476063
2016-04-06    22.180583
2016-04-07    13.194688
Name: last_seen, dtype: float64%


For the "last_seen" column, 47% of the listings had been looked at in the 2 days before the data was crawled and the dataset was created. The data is fairly even for past those two days before April 7th, 2016.

Lastly, we have the registration_year data to take a first look at.

In [139]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can immediately see something fishy about our data - that the max value is the year 9999 and the min value is the year 1000. Last I checked, time travel has not yet been discovered, so we will remove the rows of data that contain a registration year past 2016, when the data was scraped. Also, for lack of a better minumum year to use, let's go with 1885, when the automobile was invented.

In [140]:
#Removing years before the invention of the automobile, and after the year the data was scraped.

autos = autos.loc[(autos["registration_year"]>=1885) & (autos["registration_year"]<=2016),:]
autos_regyear_counts = autos["registration_year"].value_counts(normalize=True)

#Checking our results
print(autos["registration_year"].describe())
print("")
print(str(autos_regyear_counts.head(15) * 100)+"%")
print("")
print(str(autos_regyear_counts.tail(15) * 100)+"%")

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

2000    6.760781
2005    6.289497
1999    6.205951
2004    5.790364
2003    5.781796
2006    5.719672
2001    5.646837
2002    5.325507
1998    5.062017
2007    4.877788
2008    4.744971
2009    4.466485
1997    4.179431
2011    3.476789
2010    3.403954
Name: registration_year, dtype: float64%

1954    0.004284
1955    0.004284
1957    0.004284
1951    0.004284
1941    0.004284
1934    0.004284
1927    0.002142
1929    0.002142
1952    0.002142
1931    0.002142
1939    0.002142
1948    0.002142
1938    0.002142
1953    0.002142
1943    0.002142
Name: registration_year, dtype: float64%


We can see from the counts reducing between the last two .describe() calls that we eliminated another 1884 data entries by reducing the "registration_year" timeframe to ones that are actually possible. Also, our standard deviation dropped to about a 1/12th of what it was, a significant reduction in data variance that demonstrates our criteria for removing outliers was reasonable.

Next, we will look at the "brand" column, to see what the data is telling us about how the listings vary in based on the brand of automobile. Since there are 40 different brands of automobile in this data set, we are going to look at those only that have a stasticically significant share of the total number of listings. We will find a reasonable criteria for this below.

In [141]:
print(str(autos["brand"].value_counts(normalize=True).head(30) * 100)+"%")

volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
mitsubishi         0.822604
honda              0.784045
kia                0.706926
alfa_romeo         0.664082
porsche            0.612669
suzuki             0.593389
chevrolet          0.569825
chrysler           0.351321
dacia              0.263490
daihatsu           0.250637
Name: brand, dtype: float64%


Doing a bit of quick math below, we can see that the top 23 brands encompass 95.135% of our listings. We will only be omitting about 5% of the total data to eliminate 17 brands from our aggregation criteria, which is nearly half of the 40 total brands in the dataset.

In [142]:
#Demonstrating our 95% cutoff

print("The top 23 brands represent " + str(sum(autos["brand"].value_counts(normalize=True).head(23))) + "% of the dataset.")

#The brands we will be aggregating from, this list will be the number of occurences of these brands
#with index labels of the brand name
autos_top23_brands_counts = autos["brand"].value_counts().head(23)
autos_top23_brands_index = autos_top23_brands.index

#Checking our variable containing the counts of each brand in the current dataset
print("")
print(str(autos_top23_brands_counts))

The top 23 brands represent 0.9513506565840493% of the dataset.

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
Name: brand, dtype: int64


Now, we can see what will be doing the aggregation by looping through our data set by brand and adding them to a dictionary, and look at what our results tell us about the average price of a listing by the brand of the car in the listing. We will make a dictionary, the key being the brand, the value being it's average price.

avg price of listing by brand
all prices of each car of a brand / number of cars of that brand
all prices of each car of a brand  = entries of the "price" column where the brand is the current row loop element
number of cards of that brand  = finding the number of elements in the sliced selection of autos["brand"] for that brand type



In [157]:
#initializing dictionary and working variables
autos_top23_brands_mean_price = {}
sum_price = 0

#Finding the average price by brand and printing the result from the dictionary
for brand in autos_top23_brands_index:
    sum_price = autos.loc[autos["brand"] == brand,"price_usd"].sum()
    brand_count = autos["brand"].value_counts()[brand]
    autos_top23_brands_mean_price[brand] = sum_price / brand_count

#Making this dictionary a DataFrame for easy sorting of our results
autos_top23_brands_mean_price_sorted = pd.DataFrame(pd.Series(autos_top23_brands_mean_price),columns=["mean_price"])
print(autos_top23_brands_mean_price_sorted.sort_values("mean_price",ascending=False))

                  mean_price
sonstige_autos  12338.550218
mini            10613.459658
audi             9336.687454
mercedes_benz    8628.450366
bmw              8332.820518
skoda            6368.000000
kia              5982.330303
volkswagen       5402.410262
hyundai          5365.254274
toyota           5167.091062
volvo            4946.501171
nissan           4743.402525
seat             4397.230950
mazda            4112.596615
honda            4107.857923
citroen          3779.139144
ford             3749.469507
smart            3580.223903
mitsubishi       3394.572917
peugeot          3094.017229
opel             2975.241935
fiat             2813.748538
renault          2474.864607


We can see from the above some things that are predictible, such as BMW, Mercedes Benz, and Audi being some of the highest prices on average for the car by brand. Mini Cooper coming out ahead of the three of these may be the most surprising conclusion, and "sonstige_autos" being the highest average cost for it's brand.

Now, I myself was wondering what "sonstige_autos" were, having never heard of the automaker. A quick google translate search shows that sonstige simply means "other" in German, so this brand is actually a mix of various brands. Se we can disregard this brand for analysis, knowing it may be containing luxury brands that are inflating the average price.

Renault, Fiat, and Opel are the three cheapest brands by average price in that order.

But one thing to note, is that what automobiles sell for new is not necessarily a reflection of their resale value. For example: Toyotas are one of the less expensive brands when bought new, but have high resale value relative to that new sale price due to their reputation for being reliable cars that don't cost much to repair when they encounter an issue.

For the top 6 brands in terms of amount of listings (BMW, Mercedes Benz, Audi, Volkswagen, Opel, Ford). The first three are some of the most expensive, Volkswagen is somewhere in the middle, and Ford/Opel are two of the cheaper brands by listing. Let's now look further and see if there's any link between mileage and mean price for these brands.

In [169]:
#Aggregating mileage data for these six brands, starting with pulling the top 6 brands from our original dataset

autos_top6_brands = autos["brand"].value_counts().head(6)

#Using our prior work, we can select the top 6 results from the dictionary we made rather than the top 23
#and then make a dictionary for the odometer readings in the same manner as we did the mean price.

autos_top6_brands_mean_price_df = pd.DataFrame(pd.Series(autos_top23_brands_mean_price),columns=["mean_price"]).head(6)
autos_top6_brands_mean_odometer = {}

#Making dictionary of mean odometer reading for each of the top 6 brands

for brand in autos_top6_brands.index:
    sum_odometer = autos.loc[autos["brand"] == brand,"odometer_km"].sum()
    brand_count = autos["brand"].value_counts()[brand]
    autos_top6_brands_mean_odometer[brand] = sum_odometer / brand_count

#Converting the mean odometer dictionary to a Series and then adding it to our mean price DataFrame

autos_top6_brands_mean_price_df["odometer_km"] = pd.Series(autos_top6_brands_mean_odometer)

#Checking our result

autos_top6_brands_mean_price_df

mean_price    odometer_km
volkswagen     5402.410262  128707.158791
bmw            8332.820518  132572.513140
opel           2975.241935  129310.035842
mercedes_benz  8628.450366  130788.363313
audi           9336.687454  129157.386785
ford           3749.469507  124266.012872

We can see from the above that the odometer reading does not have much of an impact on the price of the car by model, as all of these brands have a similar average odometer at sale. 

We can see that the three most expensive brands of Audi, BMW, and Mercedes Benz have 3 of the 4 highest average odometer readings at sale, meaning people are still willing to pay more for these brands even when they have high odometer readings, which means brand of the automobile is the real driver of the price, rather than how much is on the vehicle's odometer.

Thank you for your time in looking at this data with me!